### PR TITLE
REV-14479 KPI Widget v2 - Enhancements

### PR DESCRIPTION
**Pull request by dfernandeza**

### CODE REVIEW

File: Lib/DateHelper.php [hunk 1]
```diff
@@ -219,9 +219,21 @@ public static function getPeriodInfo($periodName, $time = null, $timezone = 'UTC
             $range = self::getScaleRange(date('Y-m-d', strtotime('-7 days', $time)), 'week');
             $period = array($range[0], 'week', $range[1]);
             $return = array('period' => $period, 'description' => 'Last Week: ' . date('n/j/y', strtotime($range[0])) . '-' . date('n/j/y', strtotime($range[1])), 'short_description' => 'Last Week');
+        } elseif ($periodName == 'lwpp') {
+            $lw = self::getScaleRange(date('Y-m-d', strtotime('-7 days', $time)), 'week');
+            $period = array(date('Y-m-d', strtotime('-1 week', strtotime($lw[0]))), 'last_week_previous_period', date('Y-m-d', strtotime('-1 week', strtotime($lw[1]))));
```
Comment (vinaykr) [inline Lib/DateHelper.php]: Does the "lwpp" need to start from the start of week like Sunday or Monday?

File: Lib/DateHelper.php [hunk 2]
```diff
@@ -219,9 +219,21 @@ public static function getPeriodInfo($periodName, $time = null, $timezone = 'UTC
...
+            $return = Array('period' => $period, 'description' => 'Last Week Previous Period: ' . date('n/j/y', strtotime($period[0])) . ' - ' . date('n/j/y', strtotime($period[2])), 'short_description' => 'Last Week Previous Period');
         } elseif ($periodName == 'mtd') {
             $period = array(date('Y-m-01', $time), 'month', date('Y-m-d'));
             $return = array('period' => $period, 'description' => 'Month To Date: ' . date('n/j/y', strtotime($period[0])) . '-' . date('n/j/y'), 'short_description' => 'Month to Date');
+        } elseif ($periodName == 'mtdpp') {
+            $mtd = array(date('Y-m-01', $time), date('Y-m-d', $time));
+            $period = array(date('Y-m-d', strtotime('-1 month', strtotime($mtd[0]))), 'month_previous_period', date('Y-m-d', strtotime('-1 month', strtotime($mtd[1]))));
+            $return = array('period' => $period, 'description' => 'Month To Date Previous Period: ' . date('n/j/y', strtotime($period[0])) . ' - ' . date('n/j/y', strtotime($period[2])), 'short_description' => 'Month To Date Previous Period');
```
Comment (vinaykr) [inline Lib/DateHelper.php]: looks good!

File: Lib/DateHelper.php [hunk 3]
```diff
@@ -219,9 +219,21 @@ public static function getPeriodInfo($periodName, $time = null, $timezone = 'UTC
...
+        } elseif ($periodName == 'mtdpy') {
+            $mtd = array(date('Y-m-01', $time), date('Y-m-d', $time));
+            $period = array(date('Y-m-d', strtotime('-1 year', strtotime($mtd[0]))), 'month_previous_year', date('Y-m-d', strtotime('-1 year', strtotime($mtd[1]))));
+            $return = array('period' => $period, 'description' => 'Month To Date Previous Year: ' . date('n/j/y', strtotime($period[0])) . ' - ' . date('n/j/y', strtotime($period[2])), 'short_description' => 'Month to Date Previous Year');
```
Comment (vinaykr) [inline Lib/DateHelper.php]: lgtm.

File: Lib/DateHelper.php [hunk 4]
```diff
@@ -241,13 +253,21 @@ public static function getPeriodInfo($periodName, $time = null, $timezone = 'UTC
         } elseif ($periodName == 'ytd') {
             $period = array(date('Y-01-01', $time), 'year', date('Y-m-d'));
             $return = array('period' => $period, 'description' => 'Year To Date: ' . date('Y'), 'short_description' => 'Year to Date');
+        } elseif ($periodName == 'ytdpy') {
+            $ytd = array(date('Y-01-01', $time), date('Y-m-d', $time));
+            $period = array(date('Y-m-d', strtotime('-1 year', strtotime($ytd[0]))), 'year_previous_year', date('Y-m-d',strtotime('-1 year', strtotime($ytd[1]))));
+            $return =  array('period' => $period, 'description' => 'Year To Date Previous Year: ' . date('Y', strtotime('-1 year', strtotime($ytd[0]))), 'short_description' => 'Year to Date Previous Year');
```
Comment (vinaykr) [inline Lib/DateHelper.php]: lgtm.

File: Lib/DateHelper.php [hunk 5]
```diff
@@ -241,13 +253,21 @@ public static function getPeriodInfo($periodName, $time = null, $timezone = 'UTC
...
             $yearAgoTime = strtotime('-1 year', $time);
             $period = array(date('Y-01-01', $yearAgoTime), 'year', date('Y-12-31', $yearAgoTime));
             $return = array('period' => $period, 'description' => 'Last Year: ' . date('Y', $yearAgoTime), 'short_description' => 'Last Year');
         } elseif ($periodName == 'wtd') {
             $period = array(date('Y-m-d', strtotime('last sunday', $time)), 'week', date('Y-m-d'));
             $return = array('period' => $period, 'description' => 'Week To Date: ' . date('n/j/y', strtotime($period[0])) . '-' . date('n/j/y'), 'short_description' => 'Week to Date');
+        } elseif ($periodName == 'wtdpp') {
+            $wtd = date('Y-m-d', strtotime('last sunday', $time));
+            $period = array(date('Y-m-d', strtotime('-1 week', strtotime($wtd))), 'week_previous_period', date('Y-m-d', strtotime('-1 week', $time)));
+            $return = array('period' => $period, 'description' => 'Week To Date Previous Period: ' . date('n/j/y', strtotime($period[0])) . ' - ' . date('n/j/y', strtotime($period[2])), 'short_description' => 'Week to Date Previous Period');
```
Comment (vinaykr) [inline Lib/DateHelper.php]: lgtm.

File: Lib/DateHelper.php [hunk 6]
```diff
@@ -293,6 +313,18 @@ public static function getPeriodInfo($periodName, $time = null, $timezone = 'UTC
             $dayOfMonth = intval($dayOfMonth) > intval($daysInMonth) ? $daysInMonth : $dayOfMonth;
             $period = array($startDate, 'month', date('Y-m-', strtotime($startDate)) . $dayOfMonth);
             return array('period' => $period, 'description' => 'Last Month Pace: ' . date('m/d/y', strtotime($period[0])) . ' - ' . date('m/d/y', strtotime($period[2])), 'short_description' => "Last Month Pace");
+        } elseif ($periodName == 'qtdpp') {
+            $startDay = date('Y-m-d', $time);
+            $range = self::getScaleRange($startDay, 'quarter');
+            $qtd = array($range[0], date('Y-m-d', $time));
+            $period = array(date('Y-m-d',strtotime('-3 month', strtotime($qtd[0]))), 'quarter_previous_period',  date('Y-m-d',strtotime('-3 month', strtotime($qtd[1]))));
+            $return = array('period' => $period, 'description' => 'Quarter To Date Previous Period: ' . date('n/j/y', strtotime($period[0])) . ' - ' . date('n/j/y', strtotime($period[2])), 'short_description' => 'Quarter To Date Previous Period');
```
Comment (vinaykr) [inline Lib/DateHelper.php]: lgtm.

File: Lib/DateHelper.php [hunk 7]
```diff
@@ -293,6 +313,18 @@ public static function getPeriodInfo($periodName, $time = null, $timezone = 'UTC
...
+        } elseif ($periodName == 'qtdpy') {
+            $startDay = date('Y-m-d', $time);
+            $range = self::getScaleRange($startDay, 'quarter');
+            $qtd = array($range[0], date('Y-m-d', $time));
+            $period = array(date('Y-m-d',strtotime('-1 year', strtotime($qtd[0]))), 'quarter_previous_year',  date('Y-m-d',strtotime('-1 year', strtotime($qtd[1]))));
+            $return = array('period' => $period, 'description' => 'Quarter To Date Previous Year: ' . date('n/j/y', strtotime($period[0])) . ' - ' . date('n/j/y', strtotime($period[2])), 'short_description' => 'Quarter To Date Previous Year');
```
Comment (vinaykr) [inline Lib/DateHelper.php]: lgtm.

**Comment by vinaykr** [comment]
Great job!

**Comment by wayne530** [comment]
whoa, nice @dfernandeza !

**Pull request merged**